### PR TITLE
Java Client Code Review

### PR DESCRIPTION
**Pull request by nebhale**
Given some difficulties with the current implementation, I've been reviewing the existing use of the CF Java Client and Reactor to see where improvements can be made.  In pairing, a couple of serious issues were remedied and better idioms were used that should remove a couple classes of issues (primarily around what happens when the API returns 0/1/N results unexpectedly).  Overall though the code was pretty robust.

I then took an opportunity to refactor the code into a more modular design. Specifically (and I get that this might be controversial), I ensured that all calls to the Java Client were isolated in `requestXXX` methods.  This drastically enlarges the number of methods in the class, but reduces the size of any individual method.  You may also notice that these methods are also `static`.  Both of these points (small methods, `static` methods) are aiming to make any method a "pure" function with no side effects in the system.  While there's probably no real benefit in the existing code (outside of composability), it's a really good habit to get into.  It ensures that when it finally comes time to write something that really deserves this style, you'll have it as muscle memory.

I also took a look at the test code.  Specifically, there were some complaints around mocking the Java Client API.  As we had designed the API for this specific use-case, it was important to understand where the problems lay.  It was clear that as mentioned, there was a huge amount of redundant mocking code.  Going back to the primary code, you'll remember that I isolated the `requestXXX` methods.  One of the major reasons for this is that it identifies all of Java Client APIs that need to be mocked.  Conveniently, even the signatures of "mock helper" methods are described by this style as well. You'll see a direct mapping from the `requestXXX` methods to the `givenRequestXXX` with the only change being the addition of a response argument.  Looking at the changes in the tests, the I can see the argument that even these changes aren't an improvement, leaving quite a bit of code (actually more than I started with).  However, I believe that the code now is much more readable with very descriptive names (`givenRequestXXX`) and that the tests are clearer as you read through the mock setup.  That being said, since the properly mocking the tests implies that you understand how the CF server is going to behave and that isn't necessarily true for the poorly spec'd CF APIs, I would totally understand the removal of the unit tests in favor of a pure- integration test style.

As a bonus I also added a missing test (`cancel()`).

Also fixes https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/issues/88

Also fixes https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/issues/92

### CODE REVIEW

**Comment by gregturn** [comment]
What's wrong with setUp as the name Of the @Before method?

**Comment by nebhale** [comment]
Nothing particularly. It wasn't originally (it was `init`) and stylistically I avoid abbreviations. If you believe that methods should be named after what they do, then neither `setUp` nor `initialize` are very good method names.

**Comment by ericbottard** [comment]
I am generally happy with the changes direction, definitely increasing overall intent disclosure.
I am however, really not a big fan of the static approach. Nor do I buy the rationale for it : there is still some state that can be modified by side effect, so putting `static` here and there will not prevent one from doing non-functional programming.
A method that could read `doThis(someParam)` now typically reads `doThis(this.someMember, this.otherMember, this.thirdMember, someParam)` (did I mention I'm no fan of the explicit `this.` either [My IDE has syntax coloring, thank you]). This really hurts the reading flow IMHO.

When it comes to the unit tests, I definitely agree that they read better. But they're still very much disclosing the internal implementation of what they're testing, requiring careful setup of mocks to match exactly what will be called. I would lean on the side of keeping them though, maybe falling to the trap of sunk costs.

Don't get me wrong though: this is overall a big improvement. If it were me, I'd keep all of it except the `static` approach

**Comment by gregturn** [comment]
Intriguing. Spring Boot does use the explicit `this.` approach. In many places I've come to appreciate it and it's clarified intent without leaning on the IDE. 

I'm still forming my opinion on the static approach. Yes it's explicit on requiring more parameters, but since I didn't have as logged time doing this style programming I've been wiling to "try it out". 

Static methods have other impacts on programming but the most dreadful are in OOP, which don't really apply here. 

I only go with setUp because it's a swell understood semantic of a unit test to "set up the rest case". I know that it's really the @Before annotation to indicate that.

So if be willing to accept all this.

**Comment by nebhale** [comment]
@ericbottard easy stuff first.  I generally try to match house style where I can, and you'll see that the previous incarnation of the code [used `this.`](https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/blob/master/src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryTaskLauncher.java#L127) so I kept it.  That being said, I personally prefer it and I believe that it's becoming more common (as @gregturn points out Spring Boot uses it) because more and more code is being looked at in non-IDEs (e.g. GitHub) where it is helpful.  Clearly I don't care (if I was willing to match house style), so we can decide later to remove it.

As for the testing, I'm with you that they're disclosing a huge amount of internal stuff.  I think that the `launchTask` is an edge case because it does _so_ much in one go and that cancel/status are better indications.  Let's see what the LRP tests look before making a final call.

As for the `static`, I would say that the goal isn't to prevent side effects (although I'm pretty sure that's what I actually said) but rather to make it clear what players are being affected.  Maybe @poutsma can better explain my point; I seem to be struggling.

**Comment by nebhale** [comment]
@ericbottard Finished up on our work from this morning.  I even went back and dismantled the `static` usage that you weren't a fan of.  Granted that this needs to run against a real system, but I think this is pretty much ready to go.

@gregturn I've renamed the `@Before` method back to the non-descriptive `setUp()`.

**Comment by ericbottard** [comment]
Rebased and merged as fe7cc00, which includes https://github.com/ericbottard/spring-cloud-deployer-cloudfoundry/commit/75fbc13f4bafed34363a6407dabad964ee8c467a

File: src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryAppDeployer.java [hunk 1]
```diff
@@ -51,284 +51,340 @@
...
 		String deploymentId = deploymentId(request);
-		DeploymentState state = status(deploymentId).getState();
-		if (state != DeploymentState.unknown) {
-			throw new IllegalStateException(String.format("App %s is already deployed with state %s",
-				deploymentId, state));
-		}
 
-		asyncDeploy(request)
+		getStatus(deploymentId)
+			.doOnNext(status -> assertApplicationDoesNotExist(deploymentId, status))
```
Comment (nebhale) [inline src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryAppDeployer.java]: Seems like the TCK and the API Javadocs don't really match up there 😄
